Simplenote/SPAppDelegate.m

```diff
@@ -308,6 +308,48 @@ - (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(
     return YES;
 }
 
+- (void)applicationWillEnterForeground:(UIApplication *)application {
+    // Dismiss the pin lock window if the user has returned to the app before their preferred timeout length
+    if (self.pinLockWindow != nil
+        && [self.pinLockWindow isKeyWindow]
+        && [self shouldBypassPinLock]) {
+        // Bring the main window to the front, which 'dismisses' the pin lock window
+        [self.window makeKeyAndVisible];
+    }
+}
+
+- (BOOL)shouldBypassPinLock {
+    NSString *lastUsedString = [SPKeychain passwordForService:kSimplenotePasscodeServiceName account:kShareExtensionAccountName];
+    if (lastUsedString == nil) {
+        return NO;
+    }
+    
+    long lastUsedSeconds = lastUsedString.longLongValue;
+    if (lastUsedSeconds == 0) {
+        return NO;
+    }
+    
+    struct timespec ts;
+    clock_gettime(CLOCK_MONOTONIC_RAW, &ts);
+    long nowSeconds = (long)ts.tv_sec;
+    long intervalSinceLastUsed = nowSeconds - lastUsedSeconds;
+    NSInteger maxTimeoutSeconds = [self getPinLockTimeoutSeconds];
+
+    return intervalSinceLastUsed < maxTimeoutSeconds;
+}
+
+// Returns the number of seconds of the 'Timeout Lock' setting selected in SPOptionsViewController
+- (NSInteger)getPinLockTimeoutSeconds {
+    NSInteger timeoutPref = [[NSUserDefaults standardUserDefaults] integerForKey:kPinTimeoutPreferencesKey];
+    NSArray *timeoutValues = @[@0, @15, @30, @60, @120, @180, @240, @300];
+    
+    if (timeoutPref > timeoutValues.count) {
+        return 0;
+    }
+    
+    return [timeoutValues[timeoutPref] integerValue];
+}
+
```

Comment:
Nipticky Suggestion! how about moving `shouldBypassPinLock` + `getPinLockTimeoutSeconds` (and as much as possible, all related to the Pinlock!) to a class to contain all that logic?

`PinLockManager` || `PinLockController` || `LockController` || `LockManager` || `Anything better!`

Goal is to help reduce a bit the AppDelegate and contain all of the related code in a single spot.

Reply:
Haha I thought "I bet @jleandroperez will want this code to be moved to a static class" at one point ;)

Reply:
Added a new swift class for this 👍